[WIP] random: changed implementation of rand32_timer.c

decsny opened the pull request:
Changes the implementation of rand32_timer.c because it was not generating sufficiently "random"-looking values on some platforms, the implementation is now a "psuedo-xoshiro"

Requesting comments

Fixes #47908

galak commented on subsys/random/rand32_timer.c:
```diff
@@ -14,21 +14,47 @@
+	return (x << k) | (x >> (32 - k));
+}
+
+static int xoshiro128_init_state(const struct device *dev)
```
why not just initialize it when you declare it?  What's the point of this function?

decsny replied:
that was an artifact from when I was trying to do something else to get the initial state values, I have now changed it

galak commented on subsys/random/rand32_timer.c:
```diff
@@ -14,21 +14,38 @@
-static atomic_val_t _rand32_counter;
 
-#define _RAND32_INC 1000000003U
+static uint32_t state[4] = {0, 0x9b64c2b0, 0x86d3d2d4, 0xa00ae278};
```
Probably good to add a comment about where these magic numbers come from.

dleach02 replied:
@galak, these are the same magic numbers found in xoshiro.c and from the comment they are just some pseudo random values

dleach02 commented:
@d3zd3z, @ceolin, @evgeniy-paltsev 

I asked @decsny to create this draft PR to show how we can shift the timer based RNG to leverage the xoshiro algorithm to hopefully avoid the problems seen with clipped RNG values. Please review and consider this.

And note that this is for testing and platforms that do not have a good entropy source which we already warn in the build flow.

dleach02 commented on subsys/random/rand32_timer.c:
```diff
@@ -14,21 +14,38 @@
  * provide a random number generator.
  */
 
```
@decsny, pull over the attribution information from the xoshiro code into this header

evgeniy-paltsev commented on subsys/random/rand32_timer.c:
```diff
@@ -41,7 +58,7 @@ static atomic_val_t _rand32_counter;
 uint32_t z_impl_sys_rand32_get(void)
 {
-	return k_cycle_get_32() + atomic_add(&_rand32_counter, _RAND32_INC);
+	return k_cycle_get_32() + xoshiro128_next();
```
As we are now using the quality pseudo-random algorithm - let's drop this `k_cycle_get_32` which was used as an attempt of workaround the previous badly designed implementation.

d3zd3z replied:
Agree with @dleach02 here. It is import that we distinguish between "quality" of a pseudo-random algorithm for things like network timing backoffs and similar, and a cryptographic quality PRNG, which xoshiro128 is not.

evgeniy-paltsev replied:
> Agree with @dleach02 here. It is import that we distinguish between "quality" of a pseudo-random algorithm for things like network timing backoffs and similar

And I'm indeed talking about that :)

We already have multiple re-spins of rand32_timer which were trying to fix some situation when rand32_timer produce non random output (e.g see #37499) but we still have issues with it because of this k_cycle_get_32() addition - like in the current #47908 issue. It's not about cryptographic quality PRNG - it's about PRNG producing same value every call :)
Probably it' time to drop this k_cycle_get_32 call.

And as for naming - I guess it's fine to rename it to `rand32_test.c` as it anyway depends on `TEST_RANDOM_GENERATOR` option.

evgeniy-paltsev commented on subsys/random/rand32_timer.c:
```diff
@@ -14,21 +14,38 @@
-#define _RAND32_INC 1000000003U
+static uint32_t xoshiro128_next(void)
+{
+	const uint32_t result = rotl(state[0] + state[3], 7) + state[0];
+	const uint32_t t = state[1] << 9;
+
+	state[2] ^= state[0];
+	state[3] ^= state[1];
+	state[1] ^= state[2];
+	state[0] ^= state[3];
+
+	state[2] ^= t;
+	state[3] = rotl(state[3], 11);
```
I guess we want to add some locking here as we access to global data and we may call this from multiple treads. spinlock will suite perfectly fine.

dleach02 replied:
threads injecting here might actually help this test rng ;-)

Commit:
random: Change testing random generator

The old random timer test was not random-looking
enough on some platforms.

Replace with new test which is psuedo-xoshiro.

The generator is still deterministic
and does not depend on entropy at all,
but should look more random for testing.

Change name of generator tree-wide also.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

dleach02 commented:
@d3zd3z @evgeniy-paltsev @galak @ceolin can you revisit this PR?

fabiobaltieri commented:
Still a WIP?

henrikbrixandersen commented:
This is breaking `main`. I have opened a PR for reverting it for now: https://github.com/zephyrproject-rtos/zephyr/pull/53601
Perhaps this should have been rebased and tests rerun before merging?

decsny commented:
yeah this was not ready to be merged, it was a month+ old branch....

evgeniy-paltsev commented:
@decsny I'm wondering if you have any plans to rebase it and create new PR?

Thanks!
